Feat/reactions

OCannings opened the pull request.

OCannings committed:
chore(deployment): add elixir_buildpack.config

To expose some environment variables during deployment.

OCannings committed:
Merge branch 'master' of github.com:ItsJuice/juicebox

OCannings committed:
WIP webcam reaction recording

passwordinfinity committed:
feat(search): Add search component to web app

Added the (search) connected component consisting of a controlled input field
that listens for key strokes and dispatches a `search` action with what the
user types in.

passwordinfinity committed:
refactor(Search): Split component into sub-components

Search was split into three components: Search, SearchBar and ResultList.
* Search is the higher-level connected component.
* SearchBar contains an input field and listens for user input.
* ResultList updates itself and displays the current results.

OCannings committed:
refactor(Reactions): use consistent topic naming

`new.reaction` has been replaced with `reaction.sent` to make it
consistent with other topic names e.g. `video.added`

OCannings committed:
feat(Reactions): add `JuiceboxWeb.Reaction` module

The `Reaction` module uses a simple Agent to hold videos for a given
key.

* `start_link/0` Starts the Reactions process.
* `get/2` Gets a video by key.
* `put/3` Updates the video for the given key.
* `delete/2` Removes a video for a given key.

OCannings committed:
feat(Reactions): add looping reactions bar

Reactions are now shown constantly repeating at the bottom of the
screen. Users can record a new reaction to replace their previous one.
The reaction video is removed when the user disconnects.

TODO:
 * Increase test coverage
 * Replace UUID function in
   `juicebox_web/lib/web/channels/stream_channel.ex`
 * Improve reaction recording UX

passwordinfinity committed:
feat(Search): Search only if user stops typing

* Only search if it's been a while since the last time the user typed anything
into the search-bar.
* Use es6 destructuring and extration wherever possible.
* Make `result` into its own component to prepare for handling more complexity
later on.

OCannings committed:
fix(Reactions): legacy browser support

OCannings committed:
Merge branch 'feat/start-video-at-right-place' into feat/reactions

Conflicts:
	apps/juicebox_web/assets/js/routes.js

OCannings committed:
feat(Reactions): Move "reactions" to stream page

You can now react to actual videos! The `/react/:stream` route has been
removed and the reaction feature has been added to the stream page.

OCannings committed:
test(Channel): rework channel tests

The channel tests have been reworked to remove sleeps and to cover the
sending of reactions.

OCannings committed:
Merge branch 'feat/reactions' of github.com:ItsJuice/juicebox into feat/reactions

OCannings committed:
fix(Youtube): add second call for durations

passwordinfinity committed:
feat(Search): Change name of search action

Change name of the action from `search` to `receiveTerm` as the action itself
won't perform the search but let another action handle it instead.

passwordinfinity committed:
chore(thunk): add redux-thunk middleware

Redux-thunk middleware is necessary for async calls.

passwordinfinity committed:
feat(Search): Use Redux-think and fetch for async

OCannings committed:
Merge remote-tracking branch 'origin/master' into feat/reactions

Conflicts:
	apps/juicebox_web/test/lib/web/channels/stream_channel_test.exs

passwordinfinity committed:
feat(Serch): wait for search results and feed back in

OCannings committed:
feat(Search): connect search to rest of app

First working end-to-end version!

The search components have been moved into the /search directory.

OCannings committed:
Merge branch 'feat/reactions' of github.com:ItsJuice/juicebox into feat/reactions

OCannings committed:
Merge branch 'feat/add-search-fixes' into feat/reactions

Conflicts:
	apps/juicebox_web/assets/js/stores/reducers.js
	apps/juicebox_web/assets/js/videos/video-page.js

ljkbennett commented:
Hello, @OCannings! This is your first Pull Request that will be reviewed by [Ebert](https://ebertapp.io/), an automatic Code Review service. It will leave comments on this diff with potential issues and style violations found in the code as you push new commits. You can also see all the issues found on this Pull Request on its [review page](https://ebertapp.io/github/ItsJuice/juicebox/pulls/61). Please check our [documentation](https://ebertapp.io/docs) for more information.

ljkbennett commented on apps/juicebox_web/assets/js/reactions/styles.scss:
```diff
@@ -0,0 +1,21 @@
+.reaction {
+}
+
+.reaction-playing {
```
Empty selector detected ✂.

ljkbennett commented on apps/juicebox_web/assets/js/reactions/styles.scss:
```diff
@@ -0,0 +1,21 @@
+}
+
+
+.reaction {
```
Empty selector detected ✂.

ljkbennett commented on apps/juicebox_web/assets/js/reactions/styles.scss:
```diff
@@ -0,0 +1,21 @@
+    flex: 1 0 0px;
+    width: 0;
+    transition: all 0.3s;
+    &.new-reaction {
```
Rule declaration should be preceded by an empty line

[Read more about it here.](https://github.com/brigade/scss-lint/blob/master/lib/scss_lint/linter/README.md#emptylinebetweenblocks)

ljkbennett commented on apps/juicebox_web/assets/js/reactions/styles.scss:
```diff
@@ -0,0 +1,21 @@
+  video {
+    flex: 1 0 0px;
+    width: 0;
+    transition: all 0.3s;
```
`0.3` should be written without a leading zero as `.3`

[Read more about it here.](https://github.com/brigade/scss-lint/blob/master/lib/scss_lint/linter/README.md#leadingzero)

ljkbennett commented on apps/juicebox_web/assets/js/reactions/styles.scss:
```diff
@@ -0,0 +1,21 @@
+  justify-content: center;
+
+  video {
+    flex: 1 0 0px;
```
`0px` should be written without units as `0`.

ljkbennett commented on apps/juicebox_web/assets/js/reactions/styles.scss:
```diff
@@ -0,0 +1,21 @@
+.reactions {
+  width: 100%;
```
Properties should be ordered display, width

[Read more about it here.](https://github.com/brigade/scss-lint/blob/master/lib/scss_lint/linter/README.md#propertysortorder)

ljkbennett commented on apps/juicebox_web/assets/js/app.scss:
```diff
@@ -1,5 +1,4 @@
-    color: red;
-  }
+body {
+  margin: 0;
```
Properties should be ordered padding, margin

[Read more about it here.](https://github.com/brigade/scss-lint/blob/master/lib/scss_lint/linter/README.md#propertysortorder)

ljkbennett commented on apps/juicebox_web/assets/js/reactions/reducers.js:
```diff
@@ -0,0 +1,19 @@
+import { NEW_REACTION, NEXT_REACTION } from './actions';
+import { cloneDeep, omit } from 'lodash';
```
'cloneDeep' is defined but never used.

ljkbennett commented on apps/juicebox_web/assets/js/reactions/reducers.js:
```diff
@@ -0,0 +1,19 @@
+import { NEW_REACTION, NEXT_REACTION } from './actions';
```
'NEXT_REACTION' is defined but never used.

ljkbennett commented on apps/juicebox_web/assets/js/reactions/recorder.js:
```diff
@@ -0,0 +1,82 @@
+}
+
+function _errorCallback(error) {
+  console.log('navigator.getUserMedia error: ', error);
```
Unexpected console statement.

[Read more about it here.](http://eslint.org/docs/rules/no-console)

ljkbennett commented on apps/juicebox_web/assets/js/reactions/recorder.js:
```diff
@@ -0,0 +1,82 @@
+    }
+    mediaRecorder = new MediaRecorder(stream, options);
+  }else{
+    mediaRecorder = new MediaRecorder(stream);
```
Possible undefined variable: 'MediaRecorder' is not defined.

If this is intended to be a global variable, you can supress this warning by adding a  `/*global [VARIABLE NAME] */` comment on this file or set it under the `globals` section of your `.eslintrc` configuration.

[Read more about it here.](http://eslint.org/docs/rules/no-undef)

ljkbennett commented on apps/juicebox_web/assets/js/reactions/recorder.js:
```diff
@@ -0,0 +1,82 @@
+    } else if (MediaRecorder.isTypeSupported('video/webm;codecs=vp8')) {
+      var options = {mimeType: 'video/webm;codecs=vp8'};
+    }
+    mediaRecorder = new MediaRecorder(stream, options);
```
Possible undefined variable: 'MediaRecorder' is not defined.

If this is intended to be a global variable, you can supress this warning by adding a  `/*global [VARIABLE NAME] */` comment on this file or set it under the `globals` section of your `.eslintrc` configuration.

[Read more about it here.](http://eslint.org/docs/rules/no-undef)

ljkbennett commented on apps/juicebox_web/assets/js/reactions/recorder.js:
```diff
@@ -0,0 +1,82 @@
+    } else if (MediaRecorder.isTypeSupported('video/webm;codecs=vp9')) {
+      var options = {mimeType: 'video/webm;codecs=vp9'};
+    } else if (MediaRecorder.isTypeSupported('video/webm;codecs=vp8')) {
+      var options = {mimeType: 'video/webm;codecs=vp8'};
```
'options' is already defined.

ljkbennett commented on apps/juicebox_web/assets/js/reactions/recorder.js:
```diff
@@ -0,0 +1,82 @@
+      var options = {mimeType: 'video/webm;codecs=h264'};
+    } else if (MediaRecorder.isTypeSupported('video/webm;codecs=vp9')) {
+      var options = {mimeType: 'video/webm;codecs=vp9'};
+    } else if (MediaRecorder.isTypeSupported('video/webm;codecs=vp8')) {
```
Possible undefined variable: 'MediaRecorder' is not defined.

If this is intended to be a global variable, you can supress this warning by adding a  `/*global [VARIABLE NAME] */` comment on this file or set it under the `globals` section of your `.eslintrc` configuration.

[Read more about it here.](http://eslint.org/docs/rules/no-undef)

ljkbennett commented on apps/juicebox_web/assets/js/reactions/recorder.js:
```diff
@@ -0,0 +1,82 @@
+    if (MediaRecorder.isTypeSupported('video/webm;codecs=h264')) {
+      var options = {mimeType: 'video/webm;codecs=h264'};
+    } else if (MediaRecorder.isTypeSupported('video/webm;codecs=vp9')) {
+      var options = {mimeType: 'video/webm;codecs=vp9'};
```
'options' is already defined.

ljkbennett commented on apps/juicebox_web/assets/js/reactions/recorder.js:
```diff
@@ -0,0 +1,82 @@
+  if (typeof MediaRecorder.isTypeSupported == 'function'){
+    if (MediaRecorder.isTypeSupported('video/webm;codecs=h264')) {
+      var options = {mimeType: 'video/webm;codecs=h264'};
+    } else if (MediaRecorder.isTypeSupported('video/webm;codecs=vp9')) {
```
Possible undefined variable: 'MediaRecorder' is not defined.

If this is intended to be a global variable, you can supress this warning by adding a  `/*global [VARIABLE NAME] */` comment on this file or set it under the `globals` section of your `.eslintrc` configuration.

[Read more about it here.](http://eslint.org/docs/rules/no-undef)

ljkbennett commented on apps/juicebox_web/assets/js/reactions/recorder.js:
```diff
@@ -0,0 +1,82 @@
+  let mediaRecorder;
+
+  if (typeof MediaRecorder.isTypeSupported == 'function'){
+    if (MediaRecorder.isTypeSupported('video/webm;codecs=h264')) {
```
Possible undefined variable: 'MediaRecorder' is not defined.

If this is intended to be a global variable, you can supress this warning by adding a  `/*global [VARIABLE NAME] */` comment on this file or set it under the `globals` section of your `.eslintrc` configuration.

[Read more about it here.](http://eslint.org/docs/rules/no-undef)

ljkbennett commented on apps/juicebox_web/assets/js/reactions/recorder.js:
```diff
@@ -0,0 +1,82 @@
+  const encoder = new VideoEncoder();
+  let mediaRecorder;
+
+  if (typeof MediaRecorder.isTypeSupported == 'function'){
```
Possible undefined variable: 'MediaRecorder' is not defined.

If this is intended to be a global variable, you can supress this warning by adding a  `/*global [VARIABLE NAME] */` comment on this file or set it under the `globals` section of your `.eslintrc` configuration.

[Read more about it here.](http://eslint.org/docs/rules/no-undef)

ljkbennett commented on apps/juicebox_web/assets/js/reactions/recorder.js:
```diff
@@ -0,0 +1,82 @@
+import React, { Component } from 'react';
+import { connect } from 'react-redux';
```
'connect' is defined but never used.

ljkbennett commented on apps/juicebox_web/assets/js/reactions/reactions.js:
```diff
@@ -0,0 +1,40 @@
+
+class Reactions extends Component {
+  reactions() {
+    return map(this.props.reactions, (video, userId) => {
```
'reactions' is missing in props validation

littleowllabs commented:
Ebert has finished reviewing this Pull Request and has found:

* 47 possible new issues (including those that may have been commented here).
* 8 fixed issues! 🎉

But beware that this branch is **[2 commits behind the `ItsJuice:master` branch](https://github.com/ItsJuice/juicebox/compare/ItsJuice:feat/reactions...ItsJuice:master)**, and a review of an up to date branch would produce more accurate results.

You can see more details about this review at https://ebertapp.io/github/ItsJuice/juicebox/pulls/61.

<!-- Ebert auto generated comment -->

ljkbennett commented on apps/juicebox_web/assets/js/reactions/player.js:
```diff
@@ -0,0 +1,35 @@
+        className={ this.state.isNew ? 'new-reaction' : '' }
+        autoPlay
+        ref={ (video) => { this.video = video } }
+        src={ this.props.reaction }
```
'reaction' is missing in props validation

ljkbennett commented on apps/juicebox_web/assets/js/reactions/player.js:
```diff
@@ -0,0 +1,35 @@
+  }
+
+  componentWillReceiveProps({ reaction }) {
+    if (reaction !== this.props.reaction) {
```
'reaction' is missing in props validation

OCannings committed:
refactor(juicebox_web): appease Ebert

ljkbennett commented on apps/juicebox_web/assets/js/stores/main-store.prod.js:
```diff
@@ -5,7 +5,7 @@ import routes from '../routes';
 import rootReducer from './reducers';
 import { createSocket, connectToChannel, subscribeToStream } from '../sockets';
-import { queueUpdated } from '../videos/actions';
+import { queueUpdated, search } from '../videos/actions';
```
'search' is defined but never used.

ljkbennett replied:
'queueUpdated' is defined but never used.

ljkbennett commented on apps/juicebox_web/assets/js/stores/main-store.dev.js:
```diff
@@ -7,7 +7,7 @@ import thunk from 'redux-thunk';
 import { createSocket, connectToChannel, subscribeToStream } from '../sockets';
 import rootReducer from './reducers';
-import { loadInitialState, queueUpdated } from '../videos/actions';
+import { loadInitialState, queueUpdated, search } from '../videos/actions';
```
'search' is defined but never used.

ljkbennett replied:
'queueUpdated' is defined but never used.

ljkbennett replied:
'loadInitialState' is defined but never used.

ljkbennett commented on apps/juicebox_web/assets/js/search/search.js:
```diff
@@ -0,0 +1,52 @@
+    return (
+      <div>
+        <SearchBar receiveTerm={this.props.receiveTerm} />
+        <ResultList results={this.props.results} onSelect={ this.handleVideoAdded } />
```
'results' is missing in props validation

ljkbennett commented on apps/juicebox_web/assets/js/search/search.js:
```diff
@@ -0,0 +1,52 @@
+  render() {
+    return (
+      <div>
+        <SearchBar receiveTerm={this.props.receiveTerm} />
```
'receiveTerm' is missing in props validation

ljkbennett commented on apps/juicebox_web/assets/js/search/search.js:
```diff
@@ -0,0 +1,52 @@
+  handleVideoAdded(video) {
+    console.log(this.props.streamId, video);
+    this.props.addVideo({
+      streamId: this.props.streamId,
```
'streamId' is missing in props validation

ljkbennett commented on apps/juicebox_web/assets/js/search/search.js:
```diff
@@ -0,0 +1,52 @@
+
+  handleVideoAdded(video) {
+    console.log(this.props.streamId, video);
+    this.props.addVideo({
```
'addVideo' is missing in props validation

ljkbennett commented on apps/juicebox_web/assets/js/search/search.js:
```diff
@@ -0,0 +1,52 @@
+  }
+
+  handleVideoAdded(video) {
+    console.log(this.props.streamId, video);
```
'streamId' is missing in props validation

ljkbennett replied:
Unexpected console statement.

[Read more about it here.](http://eslint.org/docs/rules/no-console)

ljkbennett commented on apps/juicebox_web/assets/js/search/search-bar.js:
```diff
@@ -0,0 +1,38 @@
+  constructor(props) {
+    super(props);
+    smartSearch = smartSearch.bind(this);
+    receiveTerm = debounce(receiveTerm.bind(this), WAIT_TIME);
```
'receiveTerm' is a function.

ljkbennett commented on apps/juicebox_web/assets/js/search/search-bar.js:
```diff
@@ -0,0 +1,38 @@
+class SearchBar extends Component {
+  constructor(props) {
+    super(props);
+    smartSearch = smartSearch.bind(this);
```
'smartSearch' is a function.

ljkbennett commented on apps/juicebox_web/assets/js/search/search-bar.js:
```diff
@@ -0,0 +1,38 @@
+  receiveTerm(term);
+}
+
+function receiveTerm(term) {
```
'term' is defined but never used.

ljkbennett commented on apps/juicebox_web/assets/js/search/result.js:
```diff
@@ -0,0 +1,25 @@
+      <div className="result">
+        <li onClick={this.select}>
+          <div>{this.props.result.title}</div>
+          <div>{this.props.result.thumbnail}</div>
```
'result.thumbnail' is missing in props validation

ljkbennett replied:
'result' is missing in props validation

ljkbennett commented on apps/juicebox_web/assets/js/search/result.js:
```diff
@@ -0,0 +1,25 @@
+    return (
+      <div className="result">
+        <li onClick={this.select}>
+          <div>{this.props.result.title}</div>
```
'result.title' is missing in props validation

ljkbennett replied:
'result' is missing in props validation

ljkbennett commented on apps/juicebox_web/assets/js/search/result.js:
```diff
@@ -0,0 +1,25 @@
+  }
+
+  select() {
+    this.props.onSelect(this.props.result);
```
'result' is missing in props validation

ljkbennett replied:
'onSelect' is missing in props validation

littleowllabs commented:
Ebert has finished reviewing this Pull Request and has found:

* 56 possible new issues (including those that may have been commented here).
* 2 fixed issues! 🎉

You can see more details about this review at https://ebertapp.io/github/ItsJuice/juicebox/pulls/61.

<!-- Ebert auto generated comment -->

ljkbennett commented on apps/juicebox_web/assets/js/search/result-list.js:
```diff
@@ -0,0 +1,25 @@
+      <div className="result-list">
+        <h2>Results</h2>
+        <ul>
+          {this.results().map(result => <Result result={result} onSelect={ this.props.onSelect } />)}
```
'onSelect' is missing in props validation

ljkbennett commented on apps/juicebox_web/assets/js/search/result-list.js:
```diff
@@ -0,0 +1,25 @@
+  }
+
+  results() {
+    return this.props.results || [];
```
'results' is missing in props validation

ljkbennett commented on apps/juicebox_web/assets/js/search/actions.js:
```diff
@@ -0,0 +1,34 @@
+          dispatch(receiveResults(results));
+        });
+      }
+    }).catch(function(error) {
```
'error' is defined but never used.

ljkbennett commented on apps/juicebox_web/assets/js/reactions/recorder.js:
```diff
@@ -0,0 +1,85 @@
+  };
+}
+
+function _errorCallback(error) {
```
'error' is defined but never used.

ljkbennett commented on apps/juicebox_web/assets/js/reactions/recorder.js:
```diff
@@ -0,0 +1,85 @@
+    }
+    mediaRecorder = new MediaRecorder(stream, options);
+  }else{
+    mediaRecorder = new MediaRecorder(stream);
```
Possible undefined variable: 'MediaRecorder' is not defined.

If this is intended to be a global variable, you can supress this warning by adding a  `/*global [VARIABLE NAME] */` comment on this file or set it under the `globals` section of your `.eslintrc` configuration.

[Read more about it here.](http://eslint.org/docs/rules/no-undef)

ljkbennett commented on apps/juicebox_web/assets/js/reactions/recorder.js:
```diff
@@ -0,0 +1,85 @@
+    } else if (MediaRecorder.isTypeSupported('video/webm;codecs=vp8')) {
+      options = {mimeType: 'video/webm;codecs=vp8'};
+    }
+    mediaRecorder = new MediaRecorder(stream, options);
```
Possible undefined variable: 'MediaRecorder' is not defined.

If this is intended to be a global variable, you can supress this warning by adding a  `/*global [VARIABLE NAME] */` comment on this file or set it under the `globals` section of your `.eslintrc` configuration.

[Read more about it here.](http://eslint.org/docs/rules/no-undef)

ljkbennett commented on apps/juicebox_web/assets/js/reactions/recorder.js:
```diff
@@ -0,0 +1,85 @@
+/* global [MediaRecorder] */
```
'[MediaRecorder]' is defined but never used.

ljkbennett commented on apps/juicebox_web/assets/js/reactions/reactions.js:
```diff
@@ -0,0 +1,44 @@
+}
+
+Reactions.propTypes = {
+  reactions: PropTypes.array,
```
Possible undefined variable: 'PropTypes' is not defined.

If this is intended to be a global variable, you can supress this warning by adding a  `/*global [VARIABLE NAME] */` comment on this file or set it under the `globals` section of your `.eslintrc` configuration.

[Read more about it here.](http://eslint.org/docs/rules/no-undef)

ljkbennett commented on apps/juicebox_stream/test/support/youtube/dummy_api.ex:
```diff
@@ -1,10 +1,10 @@
 defmodule JuiceboxStream.Youtube.DummyAPI do
-  def search("bad") do
-    dummy_response(%{error: "Some API Error"})
+  def search("bad", _) do
```
Functions should have a @spec type specification.

ljkbennett commented on apps/juicebox_stream/test/support/youtube/dummy_api.ex:
```diff
@@ -1,10 +1,10 @@
 
-  def search(_) do
-    dummy_response(%{
+  def search(_, _) do
```
Functions should have a @spec type specification.

ljkbennett commented on apps/juicebox_stream/test/support/youtube/dummy_api.ex:
```diff
@@ -22,11 +21,17 @@ defmodule JuiceboxStream.Youtube.DummyAPI do
-  defp dummy_response(map) do
-    {:ok, json} = Poison.encode(map)
-    {:ok, %{body: json}}
+  def videos(_, _) do
```
Functions should have a @spec type specification.

ljkbennett commented on apps/juicebox_web/test/lib/web/channels/stream_channel_test.exs:
```diff
@@ -76,4 +76,56 @@ defmodule JuiceboxWeb.StreamChannelTest do
+    #end
+
+    test "pushes the current queue" do
+      assert_push("remote.action", %{ videos: [%{video: @video}, %{video: @video_2}], "type": "QUEUE_UPDATED" })
```
There is no whitespace around parentheses/brackets most of the time, but here there is.

ljkbennett commented on apps/juicebox_web/test/lib/web/channels/stream_channel_test.exs:
```diff
@@ -76,4 +76,56 @@ defmodule JuiceboxWeb.StreamChannelTest do
+    test "broadcasts the new reaction to all clients",
+          %{socket: socket, reaction_payload: payload, id: id} do
+
+      assert_broadcast("remote.action", %{ video: payload, user_id: id })
```
There is no whitespace around parentheses/brackets most of the time, but here there is.

ljkbennett commented on apps/juicebox_web/test/lib/web/channels/stream_channel_test.exs:
```diff
@@ -76,4 +76,56 @@ defmodule JuiceboxWeb.StreamChannelTest do
+    test "tells all clients to remove the reaction by broadcasting an empty reaction video",
+          %{socket: socket} do
+
+      assert_broadcast("remote.action", %{ video: nil, user_id: id })
```
There is no whitespace around parentheses/brackets most of the time, but here there is.

ljkbennett commented on apps/juicebox_stream/test/juicebox_stream/youtube/client_test.exs:
```diff
@@ -0,0 +1,17 @@
+        title: "Dummy",
+        description: "Video",
+        thumbnail: "test.jpg",
+        duration: 92000
```
Large numbers should be written with underscores: 92_000

ljkbennett commented on apps/juicebox_stream/test/juicebox_stream/youtube/client_test.exs:
```diff
@@ -0,0 +1,17 @@
+defmodule JuiceboxStream.Youtube.ClientTests do
+  use ExUnit.Case, async: true
+  alias JuiceboxStream.Youtube.{Client, Video}
```
Most of the time you are using the multiple single line alias/require/import/use directives but here you are using the multi-alias/require/import/use syntax

ljkbennett commented on apps/juicebox_stream/lib/juicebox_stream/youtube/time_format.ex:
```diff
@@ -0,0 +1,16 @@
+  @doc """
+  Converts a Youtube duration string to seconds (integer)
+  """
+  def parse(duration) do
```
Functions should have a @spec type specification.

ljkbennett commented on apps/juicebox_web/lib/web/channels/stream_channel.ex:
```diff
@@ -2,35 +2,61 @@ defmodule JuiceboxWeb.StreamChannel do
+    {:reply, :ok, socket}
+  end
+
+  def handle_in("reaction.sent", %{"stream_id" => stream_id, "video" => video}, socket) do
```
Functions should have a @spec type specification.

ljkbennett commented on apps/juicebox_web/lib/web/channels/stream_channel.ex:
```diff
@@ -39,9 +65,32 @@ defmodule JuiceboxWeb.StreamChannel do
     {:noreply, socket}
   end
+
+  def terminate(_reason, socket) do
```
Functions should have a @spec type specification.

ljkbennett commented on apps/juicebox_web/lib/web/channels/stream_channel.ex:
```diff
@@ -2,35 +2,61 @@ defmodule JuiceboxWeb.StreamChannel do
+  def handle_in("reaction.sent", %{"stream_id" => stream_id, "video" => video}, socket) do
+    Reactions.put(stream_id, socket.assigns.user_id, video)
+
+    broadcast_reaction(socket, %{ video: video, user_id: socket.assigns.user_id })
```
There is no whitespace around parentheses/brackets most of the time, but here there is.

littleowllabs commented:
Ebert has finished reviewing this Pull Request and has found:

* 54 possible new issues (including those that may have been commented here).
* 2 fixed issues! 🎉

You can see more details about this review at https://ebertapp.io/github/ItsJuice/juicebox/pulls/61.

<!-- Ebert auto generated comment -->

ljkbennett commented:
Hello, @OCannings! This is your first Pull Request that will be reviewed by [Ebert](https://ebertapp.io/), an automatic Code Review service. It will leave comments on this diff with potential issues and style violations found in the code as you push new commits. You can also see all the issues found on this Pull Request on its [review page](https://ebertapp.io/github/ItsJuice/juicebox/pulls/61). Please check our [documentation](https://ebertapp.io/docs) for more information.

ljkbennett commented on apps/juicebox_web/assets/js/reactions/recorder.js:
```diff
@@ -0,0 +1,85 @@
+/*global MediaRecorder*/
+
+import React, { Component, PropTypes } from 'react';
```
'PropTypes' is defined but never used.

ljkbennett commented on apps/juicebox_web/test/lib/web/channels/stream_channel_test.exs:
```diff
@@ -45,24 +45,24 @@ defmodule JuiceboxWeb.StreamChannelTest do
 
     test "after joining it receives the proper queue", %{video_2: video_2} do
-      expected_queue_response = %{ videos: [ video_2 ], type: "QUEUE_UPDATED" }
+      expected_queue_response = %{videos: [ video_2 ], type: "QUEUE_UPDATED"}
```
There is no whitespace around parentheses/brackets most of the time, but here there is.

ljkbennett commented on apps/juicebox_web/test/lib/web/channels/stream_channel_test.exs:
```diff
@@ -76,4 +76,56 @@ defmodule JuiceboxWeb.StreamChannelTest do
+    #end
+
+    test "pushes the current queue" do
+      assert_push("remote.action", %{videos: [%{video: @vide}, %{video: @video_2}], "type": "QUEUE_UPDATED" })
```
There is no whitespace around parentheses/brackets most of the time, but here there is.

ljkbennett commented on apps/juicebox_stream/lib/juicebox_stream/youtube/client.ex:
```diff
@@ -1,33 +1,52 @@
-    {:error, error}
+  defp fetch_video_meta_data(videos) do
+    Enum.map(videos, &(&1["id"]["videoId"]))
+    |> @youtube_api.videos(%{ part: "contentDetails" })
```
There is no whitespace around parentheses/brackets most of the time, but here there is.

ljkbennett commented on apps/juicebox_web/lib/juicebox/reactions_server.ex:
```diff
@@ -0,0 +1,40 @@
+  @doc """
+  Starts the Reactions process.
+  """
+  def start_link(stream_id) do
```
Functions should have a @spec type specification.

ljkbennett commented on apps/juicebox_web/lib/juicebox/reactions_server.ex:
```diff
@@ -0,0 +1,40 @@
+  @doc """
+  Gets a video by key.
+  """
+  def get(stream_id, key) do
```
Functions should have a @spec type specification.

ljkbennett commented on apps/juicebox_web/lib/juicebox/reactions_server.ex:
```diff
@@ -0,0 +1,40 @@
+  @doc """
+  Gets all videos for a given stream_id.
+  """
+  def all(stream_id) do
```
Functions should have a @spec type specification.

ljkbennett commented on apps/juicebox_web/lib/juicebox/reactions_server.ex:
```diff
@@ -0,0 +1,40 @@
+  @doc """
+  Updates the video for the given key
+  """
+  def put(stream_id, key, video) do
```
Functions should have a @spec type specification.

ljkbennett commented on apps/juicebox_web/lib/juicebox/reactions_server.ex:
```diff
@@ -0,0 +1,40 @@
+  @doc """
+  Removes a video for a given key
+  """
+  def delete(stream_id, key) do
```
Functions should have a @spec type specification.

ljkbennett commented on apps/juicebox_web/lib/juicebox/reactions_supervisor.ex:
```diff
@@ -0,0 +1,21 @@
+    Supervisor.start_link(__MODULE__, [], name: :reactions_supervisor)
+  end
+
+  def start_reactions(stream_id) do
```
Functions should have a @spec type specification.

ljkbennett commented on apps/juicebox_web/lib/juicebox/reactions_supervisor.ex:
```diff
@@ -0,0 +1,21 @@
+    Supervisor.start_child(:reactions_supervisor, [stream_id])
+  end
+
+  def init(_) do
```
Functions should have a @spec type specification.

ljkbennett commented on apps/juicebox_stream/lib/juicebox_stream/youtube/api.ex:
```diff
@@ -1,16 +1,28 @@
-      q: query,
-      key: config(:api_key)
-    })
+  def search(query, params \\ %{}) do
```
Functions should have a @spec type specification.

ljkbennett commented on apps/juicebox_stream/lib/juicebox_stream/youtube/api.ex:
```diff
@@ -1,16 +1,28 @@
+    |> parse_response
+  end
+
+  def videos(ids, params \\ %{}) do
```
Functions should have a @spec type specification.

ljkbennett commented on apps/juicebox_stream/test/juicebox_stream/youtube/time_format_test.exs:
```diff
@@ -0,0 +1,11 @@
+    assert Time.parse("PT30S") == 30
+    assert Time.parse("PT2M30S") == 150
+    assert Time.parse("PT1H2M30S") == 3750
+    assert Time.parse("P2DT1H2M30S") == 176550
```
Large numbers should be written with underscores: 176_550

littleowllabs commented:
Ebert has finished reviewing this Pull Request and has found:

* 17 possible new issues (including those that may have been commented here).
* 17 fixed issues! 🎉

You can see more details about this review at https://ebertapp.io/github/ItsJuice/juicebox/pulls/61.

<!-- Ebert auto generated comment -->

littleowllabs commented:
Ebert has finished reviewing this Pull Request and has found:

* 5 possible new issues (including those that may have been commented here).
* 19 fixed issues! 🎉

You can see more details about this review at https://ebertapp.io/github/ItsJuice/juicebox/pulls/61.

<!-- Ebert auto generated comment -->

littleowllabs commented:
Ebert has finished reviewing this Pull Request and has found:

* 2 possible new issues (including those that may have been commented here).
* 19 fixed issues! 🎉

You can see more details about this review at https://ebertapp.io/github/ItsJuice/juicebox/pulls/61.

<!-- Ebert auto generated comment -->

littleowllabs commented:
Ebert has finished reviewing this Pull Request and has found:

* 19 fixed issues! 🎉

You can see more details about this review at https://ebertapp.io/github/ItsJuice/juicebox/pulls/61.

<!-- Ebert auto generated comment -->

OCannings committed:
Merge branch 'master' into feat/reactions

Commit:
test: Don't worry, I'll fix the tests

ljkbennett commented on apps/juicebox_web/test/lib/web/channels/stream_channel_test.exs:
```diff
@@ -22,12 +23,26 @@ defmodule JuiceboxWeb.StreamChannelTest do
     {:ok, video_1: video_1, video_2: video_2}
   end
 
+  def receive_reaction %{socket: socket} do
```
Functions should have a @spec type specification.

littleowllabs commented:
Ebert has finished reviewing this Pull Request and has found:

* 2 possible new issues (including those that may have been commented here).
* 19 fixed issues! 🎉

You can see more details about this review at https://ebertapp.io/github/ItsJuice/juicebox/pulls/61.

<!-- Ebert auto generated comment -->

ljkbennett commented on apps/juicebox_web/test/lib/web/channels/stream_channel_test.exs:
```diff
@@ -22,12 +23,26 @@ defmodule JuiceboxWeb.StreamChannelTest do
+    :ok
+  end
+
+  def disconnect %{socket: socket} do
```
Functions should have a @spec type specification.

Commit:
chore: add ebert config to prevent it requiring specs

littleowllabs commented:
Ebert has finished reviewing this Pull Request and has found:

* 50 fixed issues! 🎉

You can see more details about this review at https://ebertapp.io/github/ItsJuice/juicebox/pulls/61.

<!-- Ebert auto generated comment -->

OCannings committed:
refactor(Reactions): introduce Webcam class

The Webcam class has been added to encapsulate all of the webcam related
behaviour and cross browser inconsistencies. ReactionsRecorder has been
refactored to make use of the Webcam class.

Some additional controls have been added to support previewing
reactions before sending them.

ljkbennett commented on apps/juicebox_web/assets/js/reactions/recorder.js:
```diff
@@ -0,0 +1,101 @@
+    );
+  }
+
+  webcamConnected = (streamURL) => {
```
Parsing error: Unexpected token =

ljkbennett commented:
Ebert has finished reviewing this Pull Request and has found:

* 6 possible new issues (including those that may have been commented here).
* 33 fixed issues! 🎉

You can see more details about this review at https://ebertapp.io/github/ItsJuice/juicebox/pulls/61.

<!-- Ebert auto generated comment -->
